### PR TITLE
Fix tests panic in debug

### PR DESCRIPTION
**Pull request by moCello**
Resolves: #763

### CODE REVIEW

File: src/composer.rs [hunk 1]
```diff
@@ -129,7 +129,7 @@ pub trait Composer: Sized + Index<Witness, Output = BlsScalar> {
     ) -> Witness {
         // the bits are iterated as chunks of two; hence, we require an even
         // number
-        debug_assert_eq!(num_bits & 1, 0);
+        assert_eq!(num_bits & 1, 0, "number of bits must be even");
```
Comment (vlopes11) [inline src/composer.rs:132]: Not sure if it makes sense to add production panic to check circuit development consistency. I'd rather keep it as `debug_assert_eq` and have tests checking for such case

File: src/composer.rs [hunk 2]
```diff
@@ -913,18 +919,19 @@ pub trait Composer: Sized + Index<Witness, Output = BlsScalar> {
...
     ///# Panics
-    /// This function will panic if the num_bits specified is not even, ie.
-    /// `num_bits % 2 != 0`.
+    /// This function will panic if the num_bits specified is not even since
+    /// that results in potentially incorrect circuits:
+    /// A circuit constraining 63 < 2^5 for example would validate to true.
     fn component_range(&mut self, witness: Witness, num_bits: usize) {
-        // number of bits must be even
-        debug_assert_eq!(num_bits % 2, 0);
+        assert_eq!(num_bits % 2, 0, "number of bits must be even");
```
Comment (vlopes11) [inline src/composer.rs:934]: Consider using the same assert as before: `num_bits & 1, 0`

**Review by vlopes11 — COMMENTED** [review]
Looks good overall! A comment regarding the correct change on enforcing range checks of even values

File: src/composer.rs (hunk 1 above)
Comment (ureeves) [reply src/composer.rs:132]: I have the same feeling on panicking like this, but more because it signals that we should be erroring rather than panicking. The library should validate user input against what it expects.

**Review by ureeves — CHANGES_REQUESTED** [review]
The instinct to want to actually panic on this makes sense, but I think it makes more sense to error in this situation. Here's the book on [when to panic](https://doc.rust-lang.org/book/ch09-03-to-panic-or-not-to-panic.html).

I'm open to being persuaded otherwise though.

File: src/composer.rs (hunk 1 above)
Comment (vlopes11) [reply src/composer.rs:132]: There is an important distinction tho: the validation is static and should be performed at compilation time.

Meaning: when we create a circuit implementation, we create it for a static type (i.e. `impl Circuit for Foo`). It means it could be a waste to move such check to runtime (i.e. make the function fallible & `Err`).

If we move everything to static time, then it probably makes more sense. In short: if the function receives a constant, we can just `assert` it and it would fail at compile time if the user attempts to create a circuit with wrong bits boundaries:

```rust
fn append_logic_component<BITS: usize>(
        &mut self,
        a: Witness,
        b: Witness,
        is_component_xor: bool,
    ) -> Witness {
    assert_eq!(BITS & 1, 0, "some msg");
```

The code above will not compile for invalid bits.

Comment (vlopes11) [reply src/composer.rs:132]: In fact, the problem above probably applies elsewhere. We should probably use more static stuff for other components instead of running test checks. Probably `is_component_xor` should also be a constant.

TLDR; components should receive only witnesses as arguments. Everything else is a constant

Comment (moCello) [reply src/composer.rs]: We can also circumvent any misuse if we let the user pass the amount of bit-pairs instead of bits. In the components themselves we multiply the bit-pairs by 2 and can therefore always be sure the total amount of bits is dividable by two. 

It would be more cognitive overhead on the circuit creator but would make sure the components can not be misused without the need to panic or error

Comment (moCello) [reply src/composer.rs]: I did an attempt on my above proposal combined with @vlopes11 suggestion of making the logic and range components generic over the bit (bit-halves in this case).

@vlopes11 and @ureeves please check it out under the branch: [mocello/763_fix_debug_panic_bit_pairs](https://github.com/dusk-network/plonk/tree/mocello/763_fix_debug_panic_bit_pairs) and let me know whether you like that one or the latest changes on this branch more.

Comment (ureeves) [reply src/composer.rs]: I very much like the direction of "making illegal state unrepresentable". Comes with tons of benefits. I vote for the [`bit_pairs`](https://github.com/dusk-network/plonk/tree/mocello/763_fix_debug_panic_bit_pairs) branch that you suggested.

Comment (moCello) [reply src/composer.rs]: I also like the solution with the `BIT_PAIRS` most.

Pushed to the branch, please have a final look and approve (*if* you approve that is ;) )

**Review by ureeves — APPROVED** [review]
LGTM

File: README.md [hunk 3]
```diff
@@ -50,8 +50,10 @@ impl Circuit for TestCircuit {
         composer.append_gate(constraint);
 
         // Check that a and b are in range
-        composer.component_range(a, 6);
-        composer.component_range(b, 5);
+        const HALF_SIX: usize = 3;
```
[the remark this answers is not in the document]
Comment (moCello) [reply README.md:53]: The range (and logic) component is now generic over a constant `BIT_PAIRS`. This was done so that the user has no way to construct an invalid circuit (internally we group the bits in pairs which made circuits over an uneven number of bits invalid).

In the case of the above example, we chose `3` for the generic `BIT_PAIRS`, which will result in a circuit that checks whether the witness is encoded in 6 bits or less.
To make that connection between 3 and 6 clear I chose the name `HALF_SIX`.
